### PR TITLE
Introduce `allowedNames` config for `ComposeNaming` check

### LINKED ISSUE #104: Feature Request: Allow specifying an optional pattern for `ComposableNaming`
**Is your feature request related to a problem? Please describe.**
In our architecture we can define presenter functions that return State objects. These tend to be define similarly to how [Molecule](https://github.com/cashapp/molecule) does it, where the function acts as business logic. This fails the naming check however.

```kotlin
@Composable
fun ProfilePresenter(
  userFlow: Flow<User>,
  balanceFlow: Flow<Long>,
): ProfileModel {
  val user by userFlow.collectAsState(null)
  val balance by balanceFlow.collectAsState(0L)

  return if (user == null) {
    Loading
  } else {
    Data(user.name, balance)
  }
}
```

**Describe the solution you'd like**
It would be convenient if we could have a matcher config to ignore these types.

```yml
TwitterCompose:
  ComposableNaming:
    active: false
    functionNameExcludes: ['**Presenter']
```

**Describe alternatives you've considered**
The only option right now is to suppress or disable the rule entirely.

**Additional context**
Happy to send a PR if you're open to it 👍

### PR DESCRIPTION
**Pull request by ZacSweers**
Wasn't sure what to call the editorconfig property, let me know!

Resolves #104

### CODE REVIEW

**Comment by ZacSweers** [comment]
Signed the CLA but looks like the CLA Assistant action fell over

File: rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt [hunk 1]
```diff
@@ -14,11 +15,18 @@ class ComposeNaming : ComposeKtVisitor {
     override fun visitComposable(function: KtFunction, autoCorrect: Boolean, emitter: Emitter) {
         // If it's a block we can't know if there is a return type or not from ktlint
         if (!function.hasBlockBody()) return
-        val firstLetter = function.name?.first() ?: return
+        val functionName = function.name?.takeUnless(String::isEmpty) ?: return
```
Comment (kenyee) [inline rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt:18]: can the function name ever be empty 🤔

File: rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt [hunk 2]
```diff
@@ -14,11 +15,18 @@ class ComposeNaming : ComposeKtVisitor {
...
+        val firstLetter = functionName.first()
 
         if (function.returnsValue) {
             // If it returns value, the composable should start with a lowercase letter
             if (firstLetter.isUpperCase()) {
+                // If it's allowed, we don't report it
+                val isAllowed = function.config().getSet("allowedNames", emptySet())
+                    .any {
+                        it.toRegex().matches(function.name ?: "")
```
Comment (kenyee) [inline rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt:27]: should be safe to do function.name!! here since it was validated above?

**Comment by kenyee** [comment]
> Signed the CLA but looks like the CLA Assistant action fell over

We can override..it's been a bit finicky even for me...

File: rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt (hunk 1 above)
Comment (ZacSweers) [reply rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt:18]: Tbh I'm not sure, I was thinking to just be extra careful as I have no idea what happens with local functions

File: rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt (hunk 2 above)
Comment (ZacSweers) [inline rules/common/src/main/kotlin/com/twitter/compose/rules/ComposeNaming.kt:27]:
```suggestion
                        it.toRegex().matches(functionName)
```

**Review by mrmans0n — APPROVED** [review]

File: rules/ktlint/src/test/kotlin/com/twitter/compose/rules/ktlint/ComposeNamingCheckTest.kt [hunk 3]
```diff
@@ -23,6 +24,21 @@ class ComposeNamingCheckTest {
         namingRuleAssertThat(code).hasNoLintViolations()
     }
 
+    @Test
+    fun `passes when a composable that returns values is uppercase but allowed`() {
```
Comment (mrmans0n) [inline rules/ktlint/src/test/kotlin/com/twitter/compose/rules/ktlint/ComposeNamingCheckTest.kt:28]: this test seems to be failing

Comment (ZacSweers) [reply rules/ktlint/src/test/kotlin/com/twitter/compose/rules/ktlint/ComposeNamingCheckTest.kt:28]: Fixed! I misunderstood how the naming worked. It does make the detekt name a bit wordy in context, but not sure there's an alternative to support both easily.
